src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,392 @@
+        new_width = params_.resize;
+    }
+    else {
+        new_height = params_.resize;
+        new_width = params_.resize * src.cols / src.rows;
+    }
+    try {
+        cv::resize(src, dst, cv::Size(new_width, new_height), 0, 0);
+    }
+    catch (cv::Exception &e) {
+        if (bbh_ != bad_batch_handling::skip) {
+            auto msg =
+                fmt::format("Image resize operation failed: {0:s}", e.what());
+            if (bbh_ == bad_batch_handling::error) {
+                throw std::runtime_error{msg};
+            }
+            logger::warn(msg);
+        }
+        skip_batch = true;
+    }
+}
+
+void
+image_reader::crop(cv::Mat &src,
+                   cv::Mat &dst,
+                   std::atomic_bool &skip_batch) const
+{
+    if (src.rows < params_.image_dimensions[1] ||
+        src.cols < params_.image_dimensions[2]) {
+        if (bbh_ != bad_batch_handling::skip) {
+            auto msg = fmt::format(
+                "Input image [rows: {0:n}, cols: {1:n}] is smaller than "
+                "the output image image_dimensions [rows: {2:n}, cols: {3:n}]",
+                src.rows,
+                src.cols,
+                params_.image_dimensions[1],
+                params_.image_dimensions[2]);
+            if (bbh_ == bad_batch_handling::error) {
+                throw std::invalid_argument{msg};
+            }
+            logger::warn(msg);
+        }
+        skip_batch = true;
+        return;
+    }
+
+    // center crop
+    int y = (src.rows - params_.image_dimensions[1]) / 2;
+    int x = (src.cols - params_.image_dimensions[2]) / 2;
+
+    try {
+        cv::Rect roi(
+            x, y, params_.image_dimensions[2], params_.image_dimensions[1]);
+        src(roi).copyTo(dst);
+    }
+    catch (cv::Exception &e) {
+        if (bbh_ != bad_batch_handling::skip) {
+            auto msg =
+                fmt::format("Image crop operation failed: {0:s}", e.what());
+            if (bbh_ == bad_batch_handling::error) {
+                throw std::runtime_error{msg};
+            }
+            logger::warn(msg);
+        }
+        skip_batch = true;
+    }
+}
+
+}  // namespace v1
+}  // namespace mlio
+
+#else
+
+// clang-format off
+
+#include "mlio/not_supported_error.h"
+
+namespace cv {
+class Mat {};
+}  // namespace cv
+
+namespace mlio {
+inline namespace v1 {
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmissing-noreturn"
+
+image_reader::image_reader(data_reader_params rdr_prm,
+                           image_reader_params image_prm)
+    : parallel_data_reader{std::move(rdr_prm)}, params_{std::move(image_prm)}
+{
+    bbh_ = {};
+    throw not_supported_error{"ML-IO has not been built with image reader "
+                              "support."};
+}
+
+image_reader::~image_reader() = default;
+
+intrusive_ptr<record_reader>
+image_reader::make_record_reader(data_store const &)
+{
+   throw not_supported_error{"ML-IO has not been built with image reader "
```

Comment:
For all member functions except the constructor you don't need the `not_supported_error` statement as it is impossible to construct an `image_reader` instance anyway. Instead of throwing an error, simply do nothing if the return is `void`, or return a dummy value otherwise (e.g. `return nullptr;`)